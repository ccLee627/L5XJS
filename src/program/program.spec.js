const Program = require("./index");
const Tag = require("../tag");
const Routine = require("../routine");
const Rung = require("../rung");

describe("Program Class", () => {
    describe("New Instance", () => {
        it("Accepts Proper Input", () => {
            const fn = (name, desc = null, safety = false) => () => new Program(name, desc, safety);

            expect(fn("ProgramName")).not.toThrow();
            expect(fn(null, "A Program Description")).toThrow();
            expect(fn("ProgramName", 12)).toThrow();
            expect(fn("ProgramName", "A Program Description")).not.toThrow();
            expect(fn("ProgramName", "A Program Description", 12)).toThrow();
            expect(fn("ProgramName", "A Program Description", true)).not.toThrow();
        });

        it("Initializes with Desired Document Model", () => {
            let program = new Program("ProgramName");
            expect(program).toMatchSnapshot();

            program = new Program("ProgramName", "A Description");
            expect(program).toMatchSnapshot();

            program = new Program("ProgramName", "A Description", true);
            expect(program).toMatchSnapshot();
        });
    });

    describe("Methods", () => {
        let prog;

        beforeEach(() => {
            prog = new Program("testProg");
        });

        test("_orderify: Results in the Correct Order", () => {
            prog.dom.elements.push({
                type: "element",
                name: "Tags",
                elements: []
            });

            prog.dom.elements.push({
                type: "element",
                name: "Routines",
                elements: []
            });

            prog.dom.elements.push({
                type: "element",
                name: "Description",
                elements: []
            });

            prog._orderify();

            expect(prog).toMatchSnapshot();
        });

        test("addTag: Throws on Invalid Input", () => {
            const fn = tag => () => prog.addTag(tag);

            expect(fn(new Program("notATag"))).toThrow();
            expect(fn("notATag")).toThrow();
        });

        test("addTag: Adds Tag Program Instance", () => {
            prog.addTag(new Tag("tag1", "BOOL"));
            expect(prog).toMatchSnapshot();

            prog.addTag(new Tag("tag2", "DINT"));
            expect(prog).toMatchSnapshot();

            prog.dom.elements.push({
                type: "element",
                name: "Description",
                elements: [
                    {
                        type: "cdata",
                        cdata: "A Cool Program Description"
                    }
                ]
            });

            prog.addTag(new Tag("tag3", "SINT"));
            expect(prog).toMatchSnapshot();

            const tag4 = new Tag("tag4", "REAL");
            prog.addTag(tag4);
            expect(prog).toMatchSnapshot();

            tag4.dom.attributes.Name = "not_tag4Anymore";
            expect(prog).toMatchSnapshot();
        });

        test("addRoutine: Throws on Invalid Input", () => {
            const fn = rout => () => prog.addRoutine(rout);

            expect(fn(new Program("notATag"))).toThrow();
            expect(fn("notATag")).toThrow();
        });

        test("addRoutine: Adds Routine Program Instance", () => {
            prog.addRoutine(new Routine("aRoutine"));
            expect(prog).toMatchSnapshot();

            prog.addRoutine(new Routine("bRoutine"));
            expect(prog).toMatchSnapshot();

            const rout = new Routine("cRoutine");
            prog.addRoutine(rout);
            expect(prog).toMatchSnapshot();

            rout.addRung(new Rung("XIC(someTag)NOP();"));
            expect(prog).toMatchSnapshot();
        });

        test("findTag: Throws on Invalid Input", () => {
            const fn = name => () => prog.findTag(name);

            expect(fn(new Program("notATag"))).toThrow();
            expect(fn(12)).toThrow();
            expect(fn("aTag")).not.toThrow();
        });

        test("findTag: Finds Target Tag", () => {
            prog.addTag(new Tag("aTag", "DINT"));
            prog.addTag(new Tag("bTag", "SINT"));

            const tag = new Tag("cTag", "BOOL");
            prog.addTag(tag);

            expect(prog.findTag("aTag")).not.toBeNull();
            expect(prog.findTag("aTag")).toMatchSnapshot();
            expect(prog.findTag("bTag")).not.toBeNull();
            expect(prog.findTag("bTag")).toMatchSnapshot();
            expect(prog.findTag("cTag")).not.toBeNull();
            expect(prog.findTag("cTag")).toMatchSnapshot();
            expect(prog.findTag("dTag")).toBeNull();

            prog.findTag("cTag").dom.attributes.Name = "eTag";
            expect(prog).toMatchSnapshot();
            expect(tag).toMatchSnapshot();
        });

        test("findRoutine: Throws on Invalid Input", () => {
            const fn = name => () => prog.findRoutine(name);

            expect(fn(new Program("notATag"))).toThrow();
            expect(fn(12)).toThrow();
            expect(fn("aRoutine")).not.toThrow();
        });

        test("findRoutine: Finds Target Routine", () => {
            prog.addRoutine(new Routine("aRoutine"));
            prog.addRoutine(new Routine("bRoutine"));

            const rout = new Routine("cRoutine");
            prog.addRoutine(rout);

            expect(prog.findRoutine("aRoutine")).not.toBeNull();
            expect(prog.findRoutine("aRoutine")).toMatchSnapshot();
            expect(prog.findRoutine("bRoutine")).not.toBeNull();
            expect(prog.findRoutine("bRoutine")).toMatchSnapshot();
            expect(prog.findRoutine("cRoutine")).not.toBeNull();
            expect(prog.findRoutine("cRoutine")).toMatchSnapshot();
            expect(prog.findRoutine("dRoutine")).toBeNull();

            rout.addRung(new Rung("XIC(someTag)NOP();"));
            expect(prog.findRoutine("cRoutine")).not.toBeNull();
            expect(prog.findRoutine("cRoutine")).toMatchSnapshot();
            expect(prog).toMatchSnapshot();

            prog.findRoutine("aRoutine").addRung(new Rung("XIC(anotherTag)NOP();"));
            expect(prog).toMatchSnapshot();
        });
    });

    describe("Static Methods", () => {
        test("isProgram: Returns Appropriate Judgement", () => {
            let program = new Program("ProgramName", "A Description");
            expect(Program.isProgram(program)).toBeTruthy();

            program = { notAProgram: 12 };
            expect(Program.isProgram(program)).toBeFalsy();

            program = 12;
            expect(Program.isProgram(program)).toBeFalsy();
        });
    });
});
