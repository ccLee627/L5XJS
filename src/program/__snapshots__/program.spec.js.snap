// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Program Class Methods _orderify: Results in the Correct Order 1`] = `
Program {
  "_dom": Object {
    "attributes": Object {
      "Class": "Standard",
      "Name": "testProg",
      "Type": "Normal",
    },
    "elements": Array [
      Object {
        "elements": Array [],
        "name": "Description",
        "type": "element",
      },
      Object {
        "elements": Array [],
        "name": "Tags",
        "type": "element",
      },
      Object {
        "elements": Array [],
        "name": "Routines",
        "type": "element",
      },
    ],
    "name": "Program",
    "type": "element",
  },
  "class_id": "bacac87817ce803505fcdfaf4360dc50",
  "elem_id": "33cb4159377b63ce9db11667160e39df",
}
`;

exports[`Program Class Methods addRoutine: Adds Routine Program Instance 1`] = `
Program {
  "_dom": Object {
    "attributes": Object {
      "Class": "Standard",
      "Name": "testProg",
      "Type": "Normal",
    },
    "elements": Array [
      Object {
        "elements": Array [
          Object {
            "attributes": Object {
              "Name": "aRoutine",
              "Type": "RLL",
            },
            "elements": Array [
              Object {
                "elements": Array [],
                "name": "RLLContent",
                "type": "element",
              },
            ],
            "name": "Routine",
            "type": "element",
          },
        ],
        "name": "Routines",
        "type": "element",
      },
    ],
    "name": "Program",
    "type": "element",
  },
  "class_id": "bacac87817ce803505fcdfaf4360dc50",
  "elem_id": "33cb4159377b63ce9db11667160e39df",
}
`;

exports[`Program Class Methods addRoutine: Adds Routine Program Instance 2`] = `
Program {
  "_dom": Object {
    "attributes": Object {
      "Class": "Standard",
      "Name": "testProg",
      "Type": "Normal",
    },
    "elements": Array [
      Object {
        "elements": Array [
          Object {
            "attributes": Object {
              "Name": "aRoutine",
              "Type": "RLL",
            },
            "elements": Array [
              Object {
                "elements": Array [],
                "name": "RLLContent",
                "type": "element",
              },
            ],
            "name": "Routine",
            "type": "element",
          },
          Object {
            "attributes": Object {
              "Name": "bRoutine",
              "Type": "RLL",
            },
            "elements": Array [
              Object {
                "elements": Array [],
                "name": "RLLContent",
                "type": "element",
              },
            ],
            "name": "Routine",
            "type": "element",
          },
        ],
        "name": "Routines",
        "type": "element",
      },
    ],
    "name": "Program",
    "type": "element",
  },
  "class_id": "bacac87817ce803505fcdfaf4360dc50",
  "elem_id": "33cb4159377b63ce9db11667160e39df",
}
`;

exports[`Program Class Methods addRoutine: Adds Routine Program Instance 3`] = `
Program {
  "_dom": Object {
    "attributes": Object {
      "Class": "Standard",
      "Name": "testProg",
      "Type": "Normal",
    },
    "elements": Array [
      Object {
        "elements": Array [
          Object {
            "attributes": Object {
              "Name": "aRoutine",
              "Type": "RLL",
            },
            "elements": Array [
              Object {
                "elements": Array [],
                "name": "RLLContent",
                "type": "element",
              },
            ],
            "name": "Routine",
            "type": "element",
          },
          Object {
            "attributes": Object {
              "Name": "bRoutine",
              "Type": "RLL",
            },
            "elements": Array [
              Object {
                "elements": Array [],
                "name": "RLLContent",
                "type": "element",
              },
            ],
            "name": "Routine",
            "type": "element",
          },
          Object {
            "attributes": Object {
              "Name": "cRoutine",
              "Type": "RLL",
            },
            "elements": Array [
              Object {
                "elements": Array [],
                "name": "RLLContent",
                "type": "element",
              },
            ],
            "name": "Routine",
            "type": "element",
          },
        ],
        "name": "Routines",
        "type": "element",
      },
    ],
    "name": "Program",
    "type": "element",
  },
  "class_id": "bacac87817ce803505fcdfaf4360dc50",
  "elem_id": "33cb4159377b63ce9db11667160e39df",
}
`;

exports[`Program Class Methods addRoutine: Adds Routine Program Instance 4`] = `
Program {
  "_dom": Object {
    "attributes": Object {
      "Class": "Standard",
      "Name": "testProg",
      "Type": "Normal",
    },
    "elements": Array [
      Object {
        "elements": Array [
          Object {
            "attributes": Object {
              "Name": "aRoutine",
              "Type": "RLL",
            },
            "elements": Array [
              Object {
                "elements": Array [],
                "name": "RLLContent",
                "type": "element",
              },
            ],
            "name": "Routine",
            "type": "element",
          },
          Object {
            "attributes": Object {
              "Name": "bRoutine",
              "Type": "RLL",
            },
            "elements": Array [
              Object {
                "elements": Array [],
                "name": "RLLContent",
                "type": "element",
              },
            ],
            "name": "Routine",
            "type": "element",
          },
          Object {
            "attributes": Object {
              "Name": "cRoutine",
              "Type": "RLL",
            },
            "elements": Array [
              Object {
                "elements": Array [
                  Object {
                    "attributes": Object {
                      "Number": 0,
                      "Type": "N",
                    },
                    "elements": Array [
                      Object {
                        "elements": Array [
                          Object {
                            "cdata": "XIC(someTag)NOP();",
                            "type": "cdata",
                          },
                        ],
                        "name": "Text",
                        "type": "element",
                      },
                    ],
                    "name": "Rung",
                    "type": "element",
                  },
                ],
                "name": "RLLContent",
                "type": "element",
              },
            ],
            "name": "Routine",
            "type": "element",
          },
        ],
        "name": "Routines",
        "type": "element",
      },
    ],
    "name": "Program",
    "type": "element",
  },
  "class_id": "bacac87817ce803505fcdfaf4360dc50",
  "elem_id": "33cb4159377b63ce9db11667160e39df",
}
`;

exports[`Program Class Methods addTag: Adds Tag Program Instance 1`] = `
Program {
  "_dom": Object {
    "attributes": Object {
      "Class": "Standard",
      "Name": "testProg",
      "Type": "Normal",
    },
    "elements": Array [
      Object {
        "elements": Array [
          Object {
            "attributes": Object {
              "Class": "Standard",
              "DataType": "BOOL",
              "Name": "tag1",
              "TagType": "Base",
            },
            "elements": Array [],
            "name": "Tag",
            "type": "element",
          },
        ],
        "name": "Tags",
        "type": "element",
      },
    ],
    "name": "Program",
    "type": "element",
  },
  "class_id": "bacac87817ce803505fcdfaf4360dc50",
  "elem_id": "33cb4159377b63ce9db11667160e39df",
}
`;

exports[`Program Class Methods addTag: Adds Tag Program Instance 2`] = `
Program {
  "_dom": Object {
    "attributes": Object {
      "Class": "Standard",
      "Name": "testProg",
      "Type": "Normal",
    },
    "elements": Array [
      Object {
        "elements": Array [
          Object {
            "attributes": Object {
              "Class": "Standard",
              "DataType": "BOOL",
              "Name": "tag1",
              "TagType": "Base",
            },
            "elements": Array [],
            "name": "Tag",
            "type": "element",
          },
          Object {
            "attributes": Object {
              "Class": "Standard",
              "DataType": "DINT",
              "Name": "tag2",
              "TagType": "Base",
            },
            "elements": Array [],
            "name": "Tag",
            "type": "element",
          },
        ],
        "name": "Tags",
        "type": "element",
      },
    ],
    "name": "Program",
    "type": "element",
  },
  "class_id": "bacac87817ce803505fcdfaf4360dc50",
  "elem_id": "33cb4159377b63ce9db11667160e39df",
}
`;

exports[`Program Class Methods addTag: Adds Tag Program Instance 3`] = `
Program {
  "_dom": Object {
    "attributes": Object {
      "Class": "Standard",
      "Name": "testProg",
      "Type": "Normal",
    },
    "elements": Array [
      Object {
        "elements": Array [
          Object {
            "cdata": "A Cool Program Description",
            "type": "cdata",
          },
        ],
        "name": "Description",
        "type": "element",
      },
      Object {
        "elements": Array [
          Object {
            "attributes": Object {
              "Class": "Standard",
              "DataType": "BOOL",
              "Name": "tag1",
              "TagType": "Base",
            },
            "elements": Array [],
            "name": "Tag",
            "type": "element",
          },
          Object {
            "attributes": Object {
              "Class": "Standard",
              "DataType": "DINT",
              "Name": "tag2",
              "TagType": "Base",
            },
            "elements": Array [],
            "name": "Tag",
            "type": "element",
          },
          Object {
            "attributes": Object {
              "Class": "Standard",
              "DataType": "SINT",
              "Name": "tag3",
              "TagType": "Base",
            },
            "elements": Array [],
            "name": "Tag",
            "type": "element",
          },
        ],
        "name": "Tags",
        "type": "element",
      },
    ],
    "name": "Program",
    "type": "element",
  },
  "class_id": "bacac87817ce803505fcdfaf4360dc50",
  "elem_id": "33cb4159377b63ce9db11667160e39df",
}
`;

exports[`Program Class Methods addTag: Adds Tag Program Instance 4`] = `
Program {
  "_dom": Object {
    "attributes": Object {
      "Class": "Standard",
      "Name": "testProg",
      "Type": "Normal",
    },
    "elements": Array [
      Object {
        "elements": Array [
          Object {
            "cdata": "A Cool Program Description",
            "type": "cdata",
          },
        ],
        "name": "Description",
        "type": "element",
      },
      Object {
        "elements": Array [
          Object {
            "attributes": Object {
              "Class": "Standard",
              "DataType": "BOOL",
              "Name": "tag1",
              "TagType": "Base",
            },
            "elements": Array [],
            "name": "Tag",
            "type": "element",
          },
          Object {
            "attributes": Object {
              "Class": "Standard",
              "DataType": "DINT",
              "Name": "tag2",
              "TagType": "Base",
            },
            "elements": Array [],
            "name": "Tag",
            "type": "element",
          },
          Object {
            "attributes": Object {
              "Class": "Standard",
              "DataType": "SINT",
              "Name": "tag3",
              "TagType": "Base",
            },
            "elements": Array [],
            "name": "Tag",
            "type": "element",
          },
          Object {
            "attributes": Object {
              "Class": "Standard",
              "DataType": "REAL",
              "Name": "tag4",
              "TagType": "Base",
            },
            "elements": Array [],
            "name": "Tag",
            "type": "element",
          },
        ],
        "name": "Tags",
        "type": "element",
      },
    ],
    "name": "Program",
    "type": "element",
  },
  "class_id": "bacac87817ce803505fcdfaf4360dc50",
  "elem_id": "33cb4159377b63ce9db11667160e39df",
}
`;

exports[`Program Class Methods addTag: Adds Tag Program Instance 5`] = `
Program {
  "_dom": Object {
    "attributes": Object {
      "Class": "Standard",
      "Name": "testProg",
      "Type": "Normal",
    },
    "elements": Array [
      Object {
        "elements": Array [
          Object {
            "cdata": "A Cool Program Description",
            "type": "cdata",
          },
        ],
        "name": "Description",
        "type": "element",
      },
      Object {
        "elements": Array [
          Object {
            "attributes": Object {
              "Class": "Standard",
              "DataType": "BOOL",
              "Name": "tag1",
              "TagType": "Base",
            },
            "elements": Array [],
            "name": "Tag",
            "type": "element",
          },
          Object {
            "attributes": Object {
              "Class": "Standard",
              "DataType": "DINT",
              "Name": "tag2",
              "TagType": "Base",
            },
            "elements": Array [],
            "name": "Tag",
            "type": "element",
          },
          Object {
            "attributes": Object {
              "Class": "Standard",
              "DataType": "SINT",
              "Name": "tag3",
              "TagType": "Base",
            },
            "elements": Array [],
            "name": "Tag",
            "type": "element",
          },
          Object {
            "attributes": Object {
              "Class": "Standard",
              "DataType": "REAL",
              "Name": "not_tag4Anymore",
              "TagType": "Base",
            },
            "elements": Array [],
            "name": "Tag",
            "type": "element",
          },
        ],
        "name": "Tags",
        "type": "element",
      },
    ],
    "name": "Program",
    "type": "element",
  },
  "class_id": "bacac87817ce803505fcdfaf4360dc50",
  "elem_id": "33cb4159377b63ce9db11667160e39df",
}
`;

exports[`Program Class Methods findRoutine: Finds Target Routine 1`] = `
Routine {
  "_dom": Object {
    "attributes": Object {
      "Name": "aRoutine",
      "Type": "RLL",
    },
    "elements": Array [
      Object {
        "elements": Array [],
        "name": "RLLContent",
        "type": "element",
      },
    ],
    "name": "Routine",
    "type": "element",
  },
  "class_id": "262fbaac4bd096afccf0fbf1fc161c89",
  "elem_id": "33cb4159377b63ce9db11667160e39df",
}
`;

exports[`Program Class Methods findRoutine: Finds Target Routine 2`] = `
Routine {
  "_dom": Object {
    "attributes": Object {
      "Name": "bRoutine",
      "Type": "RLL",
    },
    "elements": Array [
      Object {
        "elements": Array [],
        "name": "RLLContent",
        "type": "element",
      },
    ],
    "name": "Routine",
    "type": "element",
  },
  "class_id": "262fbaac4bd096afccf0fbf1fc161c89",
  "elem_id": "33cb4159377b63ce9db11667160e39df",
}
`;

exports[`Program Class Methods findRoutine: Finds Target Routine 3`] = `
Routine {
  "_dom": Object {
    "attributes": Object {
      "Name": "cRoutine",
      "Type": "RLL",
    },
    "elements": Array [
      Object {
        "elements": Array [],
        "name": "RLLContent",
        "type": "element",
      },
    ],
    "name": "Routine",
    "type": "element",
  },
  "class_id": "262fbaac4bd096afccf0fbf1fc161c89",
  "elem_id": "33cb4159377b63ce9db11667160e39df",
}
`;

exports[`Program Class Methods findRoutine: Finds Target Routine 4`] = `
Routine {
  "_dom": Object {
    "attributes": Object {
      "Name": "cRoutine",
      "Type": "RLL",
    },
    "elements": Array [
      Object {
        "elements": Array [
          Object {
            "attributes": Object {
              "Number": 0,
              "Type": "N",
            },
            "elements": Array [
              Object {
                "elements": Array [
                  Object {
                    "cdata": "XIC(someTag)NOP();",
                    "type": "cdata",
                  },
                ],
                "name": "Text",
                "type": "element",
              },
            ],
            "name": "Rung",
            "type": "element",
          },
        ],
        "name": "RLLContent",
        "type": "element",
      },
    ],
    "name": "Routine",
    "type": "element",
  },
  "class_id": "262fbaac4bd096afccf0fbf1fc161c89",
  "elem_id": "33cb4159377b63ce9db11667160e39df",
}
`;

exports[`Program Class Methods findRoutine: Finds Target Routine 5`] = `
Program {
  "_dom": Object {
    "attributes": Object {
      "Class": "Standard",
      "Name": "testProg",
      "Type": "Normal",
    },
    "elements": Array [
      Object {
        "elements": Array [
          Object {
            "attributes": Object {
              "Name": "aRoutine",
              "Type": "RLL",
            },
            "elements": Array [
              Object {
                "elements": Array [],
                "name": "RLLContent",
                "type": "element",
              },
            ],
            "name": "Routine",
            "type": "element",
          },
          Object {
            "attributes": Object {
              "Name": "bRoutine",
              "Type": "RLL",
            },
            "elements": Array [
              Object {
                "elements": Array [],
                "name": "RLLContent",
                "type": "element",
              },
            ],
            "name": "Routine",
            "type": "element",
          },
          Object {
            "attributes": Object {
              "Name": "cRoutine",
              "Type": "RLL",
            },
            "elements": Array [
              Object {
                "elements": Array [
                  Object {
                    "attributes": Object {
                      "Number": 0,
                      "Type": "N",
                    },
                    "elements": Array [
                      Object {
                        "elements": Array [
                          Object {
                            "cdata": "XIC(someTag)NOP();",
                            "type": "cdata",
                          },
                        ],
                        "name": "Text",
                        "type": "element",
                      },
                    ],
                    "name": "Rung",
                    "type": "element",
                  },
                ],
                "name": "RLLContent",
                "type": "element",
              },
            ],
            "name": "Routine",
            "type": "element",
          },
        ],
        "name": "Routines",
        "type": "element",
      },
    ],
    "name": "Program",
    "type": "element",
  },
  "class_id": "bacac87817ce803505fcdfaf4360dc50",
  "elem_id": "33cb4159377b63ce9db11667160e39df",
}
`;

exports[`Program Class Methods findRoutine: Finds Target Routine 6`] = `
Program {
  "_dom": Object {
    "attributes": Object {
      "Class": "Standard",
      "Name": "testProg",
      "Type": "Normal",
    },
    "elements": Array [
      Object {
        "elements": Array [
          Object {
            "attributes": Object {
              "Name": "aRoutine",
              "Type": "RLL",
            },
            "elements": Array [
              Object {
                "elements": Array [
                  Object {
                    "attributes": Object {
                      "Number": 0,
                      "Type": "N",
                    },
                    "elements": Array [
                      Object {
                        "elements": Array [
                          Object {
                            "cdata": "XIC(anotherTag)NOP();",
                            "type": "cdata",
                          },
                        ],
                        "name": "Text",
                        "type": "element",
                      },
                    ],
                    "name": "Rung",
                    "type": "element",
                  },
                ],
                "name": "RLLContent",
                "type": "element",
              },
            ],
            "name": "Routine",
            "type": "element",
          },
          Object {
            "attributes": Object {
              "Name": "bRoutine",
              "Type": "RLL",
            },
            "elements": Array [
              Object {
                "elements": Array [],
                "name": "RLLContent",
                "type": "element",
              },
            ],
            "name": "Routine",
            "type": "element",
          },
          Object {
            "attributes": Object {
              "Name": "cRoutine",
              "Type": "RLL",
            },
            "elements": Array [
              Object {
                "elements": Array [
                  Object {
                    "attributes": Object {
                      "Number": 0,
                      "Type": "N",
                    },
                    "elements": Array [
                      Object {
                        "elements": Array [
                          Object {
                            "cdata": "XIC(someTag)NOP();",
                            "type": "cdata",
                          },
                        ],
                        "name": "Text",
                        "type": "element",
                      },
                    ],
                    "name": "Rung",
                    "type": "element",
                  },
                ],
                "name": "RLLContent",
                "type": "element",
              },
            ],
            "name": "Routine",
            "type": "element",
          },
        ],
        "name": "Routines",
        "type": "element",
      },
    ],
    "name": "Program",
    "type": "element",
  },
  "class_id": "bacac87817ce803505fcdfaf4360dc50",
  "elem_id": "33cb4159377b63ce9db11667160e39df",
}
`;

exports[`Program Class Methods findTag: Finds Target Tag 1`] = `
Tag {
  "_dom": Object {
    "attributes": Object {
      "Class": "Standard",
      "DataType": "DINT",
      "Name": "aTag",
      "TagType": "Base",
    },
    "elements": Array [],
    "name": "Tag",
    "type": "element",
  },
  "class_id": "202108c6004705df7b5fd0add930619f",
  "elem_id": "33cb4159377b63ce9db11667160e39df",
}
`;

exports[`Program Class Methods findTag: Finds Target Tag 2`] = `
Tag {
  "_dom": Object {
    "attributes": Object {
      "Class": "Standard",
      "DataType": "SINT",
      "Name": "bTag",
      "TagType": "Base",
    },
    "elements": Array [],
    "name": "Tag",
    "type": "element",
  },
  "class_id": "202108c6004705df7b5fd0add930619f",
  "elem_id": "33cb4159377b63ce9db11667160e39df",
}
`;

exports[`Program Class Methods findTag: Finds Target Tag 3`] = `
Tag {
  "_dom": Object {
    "attributes": Object {
      "Class": "Standard",
      "DataType": "BOOL",
      "Name": "cTag",
      "TagType": "Base",
    },
    "elements": Array [],
    "name": "Tag",
    "type": "element",
  },
  "class_id": "202108c6004705df7b5fd0add930619f",
  "elem_id": "33cb4159377b63ce9db11667160e39df",
}
`;

exports[`Program Class Methods findTag: Finds Target Tag 4`] = `
Program {
  "_dom": Object {
    "attributes": Object {
      "Class": "Standard",
      "Name": "testProg",
      "Type": "Normal",
    },
    "elements": Array [
      Object {
        "elements": Array [
          Object {
            "attributes": Object {
              "Class": "Standard",
              "DataType": "DINT",
              "Name": "aTag",
              "TagType": "Base",
            },
            "elements": Array [],
            "name": "Tag",
            "type": "element",
          },
          Object {
            "attributes": Object {
              "Class": "Standard",
              "DataType": "SINT",
              "Name": "bTag",
              "TagType": "Base",
            },
            "elements": Array [],
            "name": "Tag",
            "type": "element",
          },
          Object {
            "attributes": Object {
              "Class": "Standard",
              "DataType": "BOOL",
              "Name": "eTag",
              "TagType": "Base",
            },
            "elements": Array [],
            "name": "Tag",
            "type": "element",
          },
        ],
        "name": "Tags",
        "type": "element",
      },
    ],
    "name": "Program",
    "type": "element",
  },
  "class_id": "bacac87817ce803505fcdfaf4360dc50",
  "elem_id": "33cb4159377b63ce9db11667160e39df",
}
`;

exports[`Program Class Methods findTag: Finds Target Tag 5`] = `
Tag {
  "_dom": Object {
    "attributes": Object {
      "Class": "Standard",
      "DataType": "BOOL",
      "Name": "eTag",
      "TagType": "Base",
    },
    "elements": Array [],
    "name": "Tag",
    "type": "element",
  },
  "class_id": "202108c6004705df7b5fd0add930619f",
  "elem_id": "33cb4159377b63ce9db11667160e39df",
}
`;

exports[`Program Class New Instance Initializes with Desired Document Model 1`] = `
Program {
  "_dom": Object {
    "attributes": Object {
      "Class": "Standard",
      "Name": "ProgramName",
      "Type": "Normal",
    },
    "elements": Array [],
    "name": "Program",
    "type": "element",
  },
  "class_id": "bacac87817ce803505fcdfaf4360dc50",
  "elem_id": "33cb4159377b63ce9db11667160e39df",
}
`;

exports[`Program Class New Instance Initializes with Desired Document Model 2`] = `
Program {
  "_dom": Object {
    "attributes": Object {
      "Class": "Standard",
      "Name": "ProgramName",
      "Type": "Normal",
    },
    "elements": Array [
      Object {
        "elements": Array [
          Object {
            "cdata": "A Description",
            "type": "cdata",
          },
        ],
        "name": "Description",
        "type": "element",
      },
    ],
    "name": "Program",
    "type": "element",
  },
  "class_id": "bacac87817ce803505fcdfaf4360dc50",
  "elem_id": "33cb4159377b63ce9db11667160e39df",
}
`;

exports[`Program Class New Instance Initializes with Desired Document Model 3`] = `
Program {
  "_dom": Object {
    "attributes": Object {
      "Class": "Safety",
      "Name": "ProgramName",
      "Type": "Normal",
    },
    "elements": Array [
      Object {
        "elements": Array [
          Object {
            "cdata": "A Description",
            "type": "cdata",
          },
        ],
        "name": "Description",
        "type": "element",
      },
    ],
    "name": "Program",
    "type": "element",
  },
  "class_id": "bacac87817ce803505fcdfaf4360dc50",
  "elem_id": "33cb4159377b63ce9db11667160e39df",
}
`;
